### PR TITLE
Generate variations dialogs

### LINKED ISSUE #7858: Add confirmation and progress modals
### Description

This issue is about adding:
- a confirmation dialog before generating all variations
- a progress dialog during the process (if needed - TBD). We should reuse existing progress dialog which appears after tapping on "Generate new variation"

### LINKED ISSUE #7860: Inform user about cases with 100+ variations
### Description

As we decided internally, we'll limit variations creation up to 100. If a user requested to create all variations for a product that would produce more than 100, we should not allow the user to continue.

This task is about informing users about such a limitation when needed.

### PR DESCRIPTION
**Pull request by wzieba**
Closes: #7858

Closes: #7860

### Description

This PR adds UI dialogs for confirmation of creating multiple variations and error cases.

#### TC1 Success
1. Select a variable product with less than 100 variations possible to generate
2. Go to variations
3. Select "Generate all variations"
4. Tap on "Ok"
5. **Assert there's a progress dialog for a few seconds and then disappears**

#### TC2 Limit exceeded
1. Select a variable product with more than 100 variations possible to generate
2. Go to variations
3. Select "Generate all variations"
4. **Assert there's an error dialog**

---

More complete test cases will be done in scope of #7877 

#### TC1

https://user-images.githubusercontent.com/5845095/202484129-24e6318a-80b3-4c88-ad39-4ff4dfadbc0f.mov

#### TC2

https://user-images.githubusercontent.com/5845095/202484230-9a3b26eb-2fb0-4cb4-bd2e-89ec79cdcbb7.mov

- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/products/VariationListViewModelTest.kt [hunk 1]
```diff
@@ -47,7 +47,8 @@ class VariationListViewModelTest : BaseUnitTest() {
                 productRepository,
                 networkStatus,
                 currencyFormatter,
-                coroutinesTestRule.testDispatchers
+                coroutinesTestRule.testDispatchers,
+                mock(),
```
Comment (wzieba) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/products/VariationListViewModelTest.kt:51]: I plan to add unit tests in scope of #7877 when we'll have the complete flow

File: WooCommerce/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -1646,6 +1646,10 @@
     <string name="variations_bulk_update_current_stock_quantity">Current stock quantity is %d</string>
     <string name="variations_bulk_update_current_stock_quantity_mixed">Current stock quantity is mixed</string>
     <string name="variations_bulk_update_stock_quantity_success">Updated stock quantity</string>
+    <string name="variations_bulk_creation_warning_title">Generation limit exceeded</string>
+    <string name="variations_bulk_creation_warning_message">Currently creation is supported for 100 variations maximum. Generating variations for this product would create %1$d variations.</string>
+    <string name="variations_bulk_creation_confirmation_title">Are you sure you want to generate all variations?</string>
+    <string name="variations_bulk_creation_confirmation_message">This will create a new variation for each and every possible combination of variation attributes (%1$d variations).</string>
```
Comment (wzieba) [inline WooCommerce/src/main/res/values/strings.xml:1655]: This message is a little modified message from the core

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListViewModel.kt [hunk 3]
```diff
@@ -271,6 +275,28 @@ class VariationListViewModel @Inject constructor(
...
+            triggerEvent(ShowGenerateVariationError.LimitExceeded(variationCandidates))
+        }
+    }
+
+    @Suppress("UnusedPrivateMember")
+    fun onGenerateVariationsConfirmed(variationCandidates: List<VariationCandidate>) {
+        viewState = viewState.copy(isProgressDialogShown = true)
+        launch {
+            @Suppress("MagicNumber")
+            delay(4000)
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListViewModel.kt:299]: Here we'll add the request to core.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListViewModel.kt [hunk 4]
```diff
@@ -271,6 +275,28 @@ class VariationListViewModel @Inject constructor(
...
+    fun onAddAllVariationsClicked() {
+        val product = viewState.parentProduct ?: return
+
+        val variationCandidates = generateVariationCandidates.invoke(product)
+
+        if (variationCandidates.size < GenerateVariationCandidates.VARIATION_CREATION_LIMIT) {
+            triggerEvent(ShowGenerateVariationConfirmation(variationCandidates))
+        } else {
+            triggerEvent(ShowGenerateVariationError.LimitExceeded(variationCandidates))
+        }
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListViewModel.kt]: I believe we don't need to send the whole list of variation candidates here. Just sending the list's size would be fine.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListViewModel.kt [hunk 5]
```diff
@@ -302,4 +328,13 @@ class VariationListViewModel @Inject constructor(
...
     object ShowBulkUpdateLimitExceededWarning : Event()
+
+    data class ShowGenerateVariationConfirmation(val variationCandidates: List<VariationCandidate>) : Event()
+
+    sealed class ShowGenerateVariationError : Event() {
+        data class LimitExceeded(val variationCandidates: List<VariationCandidate>) : ShowGenerateVariationError()
+        object NetworkError : ShowGenerateVariationError()
+    }
+
+    object ShowGenerateVariationProgress : Event()
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListViewModel.kt]: We are not using `ShowGenerateVariationProgress`, maybe we should remove this class.

File: WooCommerce/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -1646,6 +1646,10 @@
     <string name="variations_bulk_update_current_stock_quantity">Current stock quantity is %d</string>
     <string name="variations_bulk_update_current_stock_quantity_mixed">Current stock quantity is mixed</string>
     <string name="variations_bulk_update_stock_quantity_success">Updated stock quantity</string>
+    <string name="variations_bulk_creation_warning_title">Generation limit exceeded</string>
+    <string name="variations_bulk_creation_warning_message">Currently creation is supported for 100 variations maximum. Generating variations for this product would create %1$d variations.</string>
```
Comment (atorresveiga) [inline WooCommerce/src/main/res/values/strings.xml]: Maybe we can get the `100` value from `VARIATION_CREATION_LIMIT`

File: WooCommerce/src/main/res/values/strings.xml (hunk 2 above)
Comment (atorresveiga) [reply WooCommerce/src/main/res/values/strings.xml:1655]: I liked your version more 👍

**Comment by atorresveiga** [comment]
Great work @wzieba! I tested the code, and it is working as expected.
I left some minor comments, let me know what you think.

I also noted that in the image below, the title is _Generating variation_. Should we change it to  _Generating variation**s**_




### Evidence
| Confirmation  | Progress | Limit error |
| ------------- | ------------- | ------------- |
|   |   |   |

**Comment by atorresveiga** [comment]
This code is still a work in progress. Do you think we should target a `feature` branch instead of `trunk`?  This way we'll submit the code to `trunk` only when it's ready to be released.

File: WooCommerce/src/main/res/values/strings.xml (hunk 6 above)
Comment (wzieba) [reply WooCommerce/src/main/res/values/strings.xml]: Nice idea!

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListFragment.kt [hunk 7]
```diff
@@ -184,12 +201,74 @@ class VariationListFragment :
...
+            }
+            is Shown -> {
+                val dialogLabel = when (progressDialogState.cardinality) {
+                    SINGLE -> R.string.variation_create_dialog_title
+                    MULTIPLE -> R.string.variations_bulk_creation_progress_title
+                }
+                showProgressDialog(dialogLabel)
+            }
+        }
+    }
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListFragment.kt:225]: I've refactored how we handle progress dialog in this fragment as I didn't want to introduce yet another `boolean` in `ViewState`, which would also allow introducing invalid states (e.g. 2 progress bars visible).

I also changed the signature of `showProgressDialog` to not have `boolean` as `showProgressDialog(false)` doesn't look good to me and we already have the `hideProgressDialog` method.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListViewModel.kt [hunk 8]
```diff
@@ -271,6 +279,28 @@ class VariationListViewModel @Inject constructor(
...
+
+        if (variationCandidates.size < GenerateVariationCandidates.VARIATION_CREATION_LIMIT) {
+            triggerEvent(ShowGenerateVariationConfirmation(variationCandidates))
+        } else {
+            triggerEvent(ShowGenerateVariationError.LimitExceeded(variationCandidates.size))
+        }
+    }
+
+    fun onGenerateVariationsConfirmed(variationCandidates: List<VariationCandidate>) {
+        variationCandidates.size
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListViewModel.kt:295]: I added it temporarily as I had a hard time suppressing Kotlin warning which was breaking the build.

**Comment by wzieba** [comment]
> This code is still a work in progress. Do you think we should target a feature branch instead of trunk? 

I think it's okay to target `trunk` as all of this code is hidden behind a static feature flag. Eventually, it will be removed by the minification tool.

**Comment by wzieba** [comment]
Thank you @atorresveiga for the review! I've applied your suggestions, ready for the second round.

**Review by atorresveiga — APPROVED** [review]
Great work @wzieba!

> I also changed the signature of showProgressDialog to not have boolean as showProgressDialog(false) doesn't look good to me and we already have the hideProgressDialog method.

☝️ I liked this approach more

LGTM! :shipit:

**Comment by shiki** [comment]
Not sure if this is still the current version. But I just noticed that the title says "Generating variation". I think it should be plural → "Generating variations".

**Comment by wzieba** [comment]
It's fixed @shiki 👍 I think @atorresveiga spotted it somewhere.

**Comment by wzieba** [comment]
Yes: https://github.com/woocommerce/woocommerce-android/pull/7887#issuecomment-1320239897

and the commit that fixed: 7871a31

**Pull request merged**